docs/client_spec.md

```diff
@@ -0,0 +1,268 @@
+# Technical specification for writing RedisGraph client libraries
+
+By design, there is not a full standard for RedisGraph clients to adhere to. Areas such as pretty-print formatting, query validation, and transactional and multithreaded capabilities have no canonically correct behavior, and the implementer is free to choose the approach and complexity that suits them best.
+
+RedisGraph does, however, provide a compact result set format for clients that minimizes the amount of redundant data transmitted from the server. Implementers are encouraged to take advantage of this format, as it provides better performance and removes ambiguity from decoding certain data. This approach requires clients to be capable of issuing procedure calls to the server and performing a small amount of client-side caching.
+
+## Retrieving the compact result set
+
+Appending the flag `--compact` to any query issued to the GRAPH.QUERY endpoint will cause the server to issue results in the compact format. Because we don't store connection-specific configurations, all queries should be issued with this flag.
+```sh
+GRAPH.QUERY demo "MATCH (a) RETURN a" --compact
+```
+
+## Formatting differences in the compact result set
+
+The result set has the same overall structure as described in the [Result Set documentation](result_structure.md#top-level-members).
+
+Certain values are emitted as integer IDs rather than strings:
+
+1. Node labels
+2. Relationship types
+3. Property keys
+
+Instructions on how to efficiently convert these IDs in the [Procedure Calls](#procedure-calls) section below.
+
+Additionally, two enums are exposed:
+
+[ColumnType](https://github.com/RedisGraph/RedisGraph/blob/ff108d7e21061025166a35d29be1a1cb5bac6d55/src/resultset/formatters/resultset_formatter.h#L14-L19) indicates what type of value is held in each column (more formally, that offset into each row of the result set). Each entry in the header row will be a 2-array, with this enum in the first position and the column name string in the second.
+
+[PropertyType](https://github.com/RedisGraph/RedisGraph/blob/ff108d7e21061025166a35d29be1a1cb5bac6d55/src/resultset/formatters/resultset_formatter.h#L21-L28) indicates the data type (such as integer or string) of each returned scalar value. Each scalar values is emitted as a 2-array, with this enum in the first position and the actual value in th second. A column can consist exclusively of scalar values, such as both of the columns created by `RETURN a.value, 'this literal string'`. Each property on a graph entity also has a scalar as its value, so this construction is nested in each value of the properties array when a column contains a node or relationship.
```

Comment:
`th second`  -> the

Reply:
Ugh, good catches! Sometimes I miss text editors that have spell check ;)